libcalico-go/lib/epstatusfile/status_file_watcher.go

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2025 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package epstatusfile
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+)
+
+// FileWatcher monitors a directory and reports events (creation, update and deletion) on file updates.
+// It tries to use fsnotify to watch the directory. If fsnotify fails, it switches to polling mode.
+
+type Callbacks struct {
+	OnFileCreation func(fileName string)
+	OnFileUpdate   func(fileName string)
+	OnFileDeletion func(fileName string)
+}
+
+type FileWatcher struct {
+	dir                 string
+	lastState           map[string]os.FileInfo
+	mu                  sync.Mutex
+	stopChan            chan struct{}
+	fsWatcher           *fsnotify.Watcher
+	pollIntervalSeconds int
+
+	callbacks Callbacks
+}
+
+// NewWatcher creates a new Watcher instance.
+func NewFileWatcher(dir string, pollIntervalSeconds int) *FileWatcher {
+	return &FileWatcher{
+		dir:                 dir,
+		lastState:           make(map[string]os.FileInfo),
+		stopChan:            make(chan struct{}),
+		pollIntervalSeconds: pollIntervalSeconds,
+	}
+}
+
+func (w *FileWatcher) SetCallbacks(callbacks Callbacks) {
+	w.callbacks = callbacks
+}
+
+func (w *FileWatcher) Start() {
+	go w.runWatcher()
+}
+
+// Start begins watching the directory.
+func (w *FileWatcher) runWatcher() {
+	var err error
+
+	// Get current state of the directory and emit initial events.
+	w.checkForChanges()
+
+	w.fsWatcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		log.WithError(err).Error("Error initializing fsnotify. Falling back to polling.")
+		go w.pollDirectory()
+		return
+	}
+	defer w.fsWatcher.Close()
+
+	err = w.fsWatcher.Add(w.dir)
+	if err != nil {
+		log.WithError(err).Error("Error adding directory to fsnotify. Falling back to polling.")
+		go w.pollDirectory()
+		return
+	}
+
+	log.WithField("dir", w.dir).Info("Started watching directory.")
+
+	// Listen for events
+	for {
+		select {
+		case event, ok := <-w.fsWatcher.Events:
+			if !ok {
+				// Stop if channel is closed.
```

Comment:
Is this expected?  I'd have thought you'd want to recreate the watch in this case.